File: backend/src/engine/dialogue_split.py
"""LLM NPC 응답을 UI용 화자 블록으로 나눔 — 멘트(대사) 단위 카드."""

from __future__ import annotations

import re

from .story_facts import is_invalid_speaker_label

MAX_DISPLAY_SEGMENTS = 5
MAX_NARRATION_BLOCKS = 1

_HR_LINE = re.compile(r"^---+\s*$|^\*{3}\s*$|^_{3,}\s*$")
_DIALOGUE_START = re.compile(r'^["\'…「]|^\.\.\.')
_ACTION_START = re.compile(r"^\([^)]*\)")
_SPEAKER_HEADER = re.compile(
    r"^\*{0,2}([^*\n]{2,40}?)\*{0,2}\s*(?:\([^)]+\)|[:：])"
)


def normalize_assistant_text(text: str) -> str:
    """`---` 구분선 제거, 과도한 빈 줄 축소."""
    t = (text or "").strip()
    if not t:
        return ""
    t = re.sub(r"\n\s*---+\s*\n", "\n\n", t)
    t = re.sub(r"(?:^\s*---+\s*\n|\n\s*---+\s*$)", "", t)
    t = re.sub(r"\n{3,}", "\n\n", t)
    return t.strip()


def compact_assistant_segments(
    segments: list[dict[str, str]],
    *,
    max_total: int = MAX_DISPLAY_SEGMENTS,
    max_narration_blocks: int = MAX_NARRATION_BLOCKS,
) -> list[dict[str, str]]:
    """내레이션만 합치고, NPC 멘트 카드는 유지. 상한 초과 시 뒤쪽부터 합침."""
    if not segments:
        return []

    merged: list[dict[str, str]] = []
    for seg in segments:
        speaker = seg.get("speaker", "")
        text = str(seg.get("text", "")).strip()
        if not text:
            continue
        if (
            merged
            and speaker == "내레이션"
            and merged[-1].get("speaker") == "내레이션"
        ):
            merged[-1]["text"] = merged[-1]["text"] + "\n\n" + text
        else:
            merged.append({"speaker": speaker, "text": text})

    narr_indices = [i for i, s in enumerate(merged) if s.get("speaker") == "내레이션"]
    if len(narr_indices) > max_narration_blocks:
        all_narr = "\n\n".join(merged[i]["text"] for i in narr_indices)
        out: list[dict[str, str]] = []
        inserted = False
        for i, seg in enumerate(merged):
            if seg.get("speaker") == "내레이션":
                if not inserted and i == narr_indices[0]:
                    out.append({"speaker": "내레이션", "text": all_narr})
                    inserted = True
                continue
            out.append(seg)
        merged = out

    if len(merged) <= max_total:
        return merged

    narr_only = [s for s in merged if s.get("speaker") == "내레이션"]
    npc_only = [s for s in merged if s.get("speaker") != "내레이션"]
    narr_slot = 1 if narr_only else 0
    keep_npcs = max(1, max_total - narr_slot)

    if len(npc_only) <= keep_npcs:
        return (npc_only + narr_only)[:max_total]

    kept = npc_only[: keep_npcs - 1]
    overflow = npc_only[keep_npcs - 1 :]
    overflow_text = "\n\n".join(s["text"] for s in overflow)
    kept.append({"speaker": overflow[0]["speaker"], "text": overflow_text})
    if narr_only:
        kept.append({"speaker": "내레이션", "text": narr_only[0]["text"]})
    return kept[:max_total]


def _speaker_header_name(line: str) -> str | None:
    """`이름 (행동)` / `**이름** (행동)` / `이름:` 형태에서 이름 추출."""
    stripped = line.lstrip()
    if not stripped or _is_dialogue_line(stripped):
        return None
    clean = stripped.replace("**", "").strip()
    m = _SPEAKER_HEADER.match(clean) or _SPEAKER_HEADER.match(stripped)
    if m:
        cand = m.group(1).strip()
        if cand and not _is_dialogue_line(cand):
            return cand
    bold = re.match(r"^\*\*([^*]+)\*\*", stripped)
    if bold:
        cand = bold.group(1).strip()
        if cand:
            rest = stripped[bold.end() :].strip()
            if not rest or rest[0] == "(":
                return cand
    return None


def _resolve_speaker_label(label: str, names: list[str]) -> str:
    for name in names:
        if label == name or label.startswith(name) or name.startswith(label):
            return name
    return label


def _is_player_speaker(label: str, player_name: str | None) -> bool:
    if not player_name or not label:
        return False
    p = player_name.strip()
    n = label.strip()
    return n == p or n.startswith(p) or p.startswith(n)


def _filter_speaker(
    label: str, names: list[str], *, player_name: str | None = None
) -> str | None:
    """NPC 카드 화자로 쓸 수 있으면 정규화된 이름, 아니면 None."""
    resolved = _resolve_speaker_label(label, names)
    if _is_player_speaker(resolved, player_name) or _is_player_speaker(label, player_name):
        return None
    if is_invalid_speaker_label(resolved):
        return None
    return resolved


def _discover_speaker_names(
    text: str, known: list[str], *, player_name: str | None = None
) -> list[str]:
    found: set[str] = set()
    for n in known:
        if n and _filter_speaker(str(n), known, player_name=player_name):
            found.add(_resolve_speaker_label(str(n), known))
    for line in text.split("\n"):
        cand = _speaker_header_name(line.strip())
        if cand:
            acc = _filter_speaker(cand, list(found) + known, player_name=player_name)
            if acc:
                found.add(acc)
    return sorted(found, key=len, reverse=True)


def _preprocess_speaker_line_breaks(text: str, names: list[str]) -> str:
    """빈 줄 없이 `이름 (행동)` 만 줄바꿈된 LLM 출력도 블록 경계로 삼음."""
    lines = text.split("\n")
    out: list[str] = []
    for line in lines:
        stripped = line.strip()
        if out and stripped and _speaker_header_name(stripped):
            if out[-1].strip():
                out.append("")
        out.append(line)
    return "\n".join(out)


def _speaker_for_first_line(
    first_line: str, names: list[str], *, player_name: str | None = None
) -> str | None:
    stripped = first_line.lstrip()
    header = _speaker_header_name(stripped)
    if header:
        return _filter_speaker(header, names, player_name=player_name)
    bold = re.match(r"^\*\*([^*]+)\*\*", stripped)
    if bold:
        label = bold.group(1).strip()
        return _filter_speaker(label, names, player_name=player_name)
    for name in names:
        if stripped.startswith(name):
            rest = stripped[len(name) :]
            if not rest or rest[0] in " \t(":
                return _filter_speaker(name, names, player_name=player_name)
    return None


def _speaker_for_block(
    block: str, names: list[str], *, player_name: str | None = None
) -> str | None:
    first = block.split("\n", 1)[0]
    return _speaker_for_first_line(first, names, player_name=player_name)


def _first_line(block: str) -> str:
    return block.strip().split("\n", 1)[0].strip()


def _is_dialogue_line(line: str) -> bool:
    return bool(_DIALOGUE_START.match(line))


def _is_action_line(line: str) -> bool:
    return bool(_ACTION_START.match(line))


def _block_has_dialogue(block: str) -> bool:
    for line in block.split("\n"):
        if _is_dialogue_line(line.strip()):
            return True
    return False


def _strip_npc_prefix(line: str, names: list[str]) -> str:
    stripped = line.lstrip()
    bold = re.match(r"^\*\*([^*]+)\*\*", stripped)
    if bold:
        return re.sub(r"^\*\*[^*]+\*\*", "", stripped).strip()
    for name in names:
        if stripped.startswith(name):
            return stripped[len(name) :].strip()
    return stripped


def _is_npc_header_only(block: str, names: list[str]) -> bool:
    """`**이름** (행동)` 만 있고 대사가 없을 때만 True."""
    if _speaker_for_block(block, names) is None:
        return False
    if _block_has_dialogue(block):
        return False
    lines = [ln.strip() for ln in block.split("\n") if ln.strip()]
    if len(lines) != 1:
        return False
    rest = _strip_npc_prefix(lines[0], names)
    rest = re.sub(r"^\([^)]*\)", "", rest).strip()
    return not rest


def _append_narration(out: list[dict[str, str]], block: str) -> None:
    if out and out[-1].get("speaker") == "내레이션":
        out[-1]["text"] = out[-1]["text"] + "\n\n" + block
    else:
        out.append({"speaker": "내레이션", "text": block})


def _split_by_utterances(
    blocks: list[str],
    names: list[str],
    *,
    player_name: str | None = None,
) -> list[dict[str, str]]:
    """멘트(따옴표 대사)마다 카드 1개. 직전 (행동)·NPC 헤더는 해당 멘트에 붙임."""
    out: list[dict[str, str]] = []
    active_npc: str | None = None
    pending_prefix = ""

    def flush_prefix_to_last() -> None:
        nonlocal pending_prefix
        if not pending_prefix:
            return
        if out and out[-1]["speaker"] != "내레이션":
            out[-1]["text"] = out[-1]["text"] + "\n\n" + pending_prefix
        elif active_npc:
            out.append({"speaker": active_npc, "text": pending_prefix})
        else:
            out.append({"speaker": "내레이션", "text": pending_prefix})
        pending_prefix = ""

    for block in blocks:
        if not block or _HR_LINE.match(block):
            continue

        first = _first_line(block)
        header_name = _speaker_header_name(first)
        if header_name and _filter_speaker(header_name, names, player_name=player_name) is None:
            flush_prefix_to_last()
            active_npc = None
            _append_narration(out, block)
            continue

        named = _speaker_for_block(block, names, player_name=player_name)

        if named:
            active_npc = named
            if _is_npc_header_only(block, names):
                pending_prefix = (
                    f"{pending_prefix}\n\n{block}" if pending_prefix else block
                )
                continue
            text = f"{pending_prefix}\n\n{block}" if pending_prefix else block
            out.append({"speaker": named, "text": text})
            pending_prefix = ""
            continue

        if _is_action_line(first) and not _block_has_dialogue(block):
            pending_prefix = (
                f"{pending_prefix}\n\n{block}" if pending_prefix else block
            )
            continue

        if _is_dialogue_line(first) or _block_has_dialogue(block):
            speaker = active_npc or "내레이션"
            text = f"{pending_prefix}\n\n{block}" if pending_prefix else block
            out.append({"speaker": speaker, "text": text})
            pending_prefix = ""
            continue

        flush_prefix_to_last()
        active_npc = None
        _append_narration(out, block)

    flush_prefix_to_last()
    return out


def split_assistant_into_segments(
    text: str,
    npc_names: list[str],
    *,
    player_name: str | None = None,
) -> list[dict[str, str]]:
    """`text`를 멘트 단위 화자 블록으로 분리."""
    text = normalize_assistant_text(text)
    if not text:
        return []

    known = [str(n).strip() for n in npc_names if n and str(n).strip()]
    names = _discover_speaker_names(text, known, player_name=player_name)
    if not names:
        return [{"speaker": "응답", "text": text}]

    text = _preprocess_speaker_line_breaks(text, names)
    raw_blocks = re.split(r"\n\s*\n+", text)
    blocks = [b.strip() for b in raw_blocks if b.strip() and not _HR_LINE.match(b.strip())]

    utterances = _split_by_utterances(blocks, names, player_name=player_name)
    return compact_assistant_segments(utterances)

File: frontend/src/utils/dialogueSplit.ts
import type { NpcSegment } from "../api/play";

const MAX_DISPLAY_SEGMENTS = 5;
const MAX_NARRATION_BLOCKS = 1;

const HR_LINE = /^---+\s*$|^\*{3}\s*$|^_{3,}\s*$/;
const DIALOGUE_START = /^["'…「]|^\.\.\./;
const ACTION_START = /^\([^)]*\)/;
const SPEAKER_HEADER = /^\*{0,2}([^*\n]{2,40}?)\*{0,2}\s*(?:\([^)]+\)|[:：])/;

const INVALID_SPEAKER_MARKERS = [
  "팟:",
  "플랍",
  "턴:",
  "턴 ",
  "리버",
  "보드",
  "차례",
  "카톡",
  "문자",
  "위치:",
  "핸드",
  "♠",
  "♥",
  "♦",
  "♣",
  " — ",
];

function isInvalidSpeakerLabel(name: string): boolean {
  const n = (name || "").trim();
  if (!n) return true;
  if (n === "내레이션" || n === "응답" || n === "모르는 번호") return false;
  if (INVALID_SPEAKER_MARKERS.some((m) => n.includes(m))) return true;
  if (/[♠♥♦♣]/.test(n)) return true;
  if (/^\d/.test(n)) return true;
  if (n.length > 36) return true;
  return false;
}

function isPlayerSpeaker(label: string, playerName: string | null | undefined): boolean {
  if (!playerName || !label) return false;
  const p = playerName.trim();
  const n = label.trim();
  return n === p || n.startsWith(p) || p.startsWith(n);
}

function filterSpeaker(
  label: string,
  names: string[],
  playerName?: string | null,
): string | null {
  const resolved = resolveSpeakerLabel(label, names);
  if (isPlayerSpeaker(resolved, playerName) || isPlayerSpeaker(label, playerName)) return null;
  if (isInvalidSpeakerLabel(resolved)) return null;
  return resolved;
}

export function normalizeAssistantText(text: string): string {
  let t = (text || "").trim();
  if (!t) return "";
  t = t.replace(/\n\s*---+\s*\n/g, "\n\n");
  t = t.replace(/(?:^\s*---+\s*\n|\n\s*---+\s*$)/g, "");
  t = t.replace(/\n{3,}/g, "\n\n");
  return t.trim();
}

function speakerHeaderName(line: string): string | null {
  const stripped = line.replace(/^\s+/, "");
  if (!stripped || isDialogueLine(stripped)) return null;
  const clean = stripped.replace(/\*\*/g, "").trim();
  const m = SPEAKER_HEADER.exec(clean) ?? SPEAKER_HEADER.exec(stripped);
  if (m) {
    const cand = m[1].trim();
    if (cand && !isDialogueLine(cand)) return cand;
  }
  const bold = stripped.match(/^\*\*([^*]+)\*\*/);
  if (bold) {
    const cand = bold[1].trim();
    if (cand) {
      const rest = stripped.slice(bold[0].length).trim();
      if (!rest || rest[0] === "(") return cand;
    }
  }
  return null;
}

function resolveSpeakerLabel(label: string, names: string[]): string {
  for (const name of names) {
    if (label === name || label.startsWith(name) || name.startsWith(label)) return name;
  }
  return label;
}

function discoverSpeakerNames(
  text: string,
  known: string[],
  playerName?: string | null,
): string[] {
  const found = new Set<string>();
  for (const n of known) {
    if (n && filterSpeaker(n, known, playerName)) {
      found.add(resolveSpeakerLabel(n, known));
    }
  }
  for (const line of text.split("\n")) {
    const cand = speakerHeaderName(line.trim());
    if (cand) {
      const acc = filterSpeaker(cand, [...found, ...known], playerName);
      if (acc) found.add(acc);
    }
  }
  return Array.from(found).sort((a, b) => b.length - a.length);
}

function preprocessSpeakerLineBreaks(text: string): string {
  const lines = text.split("\n");
  const out: string[] = [];
  for (const line of lines) {
    const stripped = line.trim();
    if (out.length && stripped && speakerHeaderName(stripped)) {
      if (out[out.length - 1].trim()) out.push("");
    }
    out.push(line);
  }
  return out.join("\n");
}

function speakerForFirstLine(
  firstLine: string,
  names: string[],
  playerName?: string | null,
): string | null {
  const stripped = firstLine.replace(/^\s+/, "");
  const header = speakerHeaderName(stripped);
  if (header) return filterSpeaker(header, names, playerName);
  const bold = stripped.match(/^\*\*([^*]+)\*\*/);
  if (bold) return filterSpeaker(bold[1].trim(), names, playerName);
  for (const name of names) {
    if (stripped.startsWith(name)) {
      const rest = stripped.slice(name.length);
      if (rest.length === 0 || rest[0] === " " || rest[0] === "\t" || rest[0] === "(") {
        return filterSpeaker(name, names, playerName);
      }
    }
  }
  return null;
}

function speakerForBlock(block: string, names: string[], playerName?: string | null): string | null {
  const first = block.split("\n", 1)[0];
  return speakerForFirstLine(first, names, playerName);
}

function firstLine(block: string): string {
  return block.trim().split("\n", 1)[0].trim();
}

function isDialogueLine(line: string): boolean {
  return DIALOGUE_START.test(line);
}

function isActionLine(line: string): boolean {
  return ACTION_START.test(line);
}

function blockHasDialogue(block: string): boolean {
  for (const line of block.split("\n")) {
    if (isDialogueLine(line.trim())) return true;
  }
  return false;
}

function stripNpcPrefix(line: string, names: string[]): string {
  const stripped = line.replace(/^\s+/, "");
  const bold = stripped.match(/^\*\*([^*]+)\*\*/);
  if (bold) return stripped.replace(/^\*\*[^*]+\*\*/, "").trim();
  for (const name of names) {
    if (stripped.startsWith(name)) return stripped.slice(name.length).trim();
  }
  return stripped;
}

function isNpcHeaderOnly(block: string, names: string[]): boolean {
  if (speakerForBlock(block, names) === null || blockHasDialogue(block)) return false;
  const lines = block.split("\n").map((l) => l.trim()).filter(Boolean);
  if (lines.length !== 1) return false;
  let rest = stripNpcPrefix(lines[0], names);
  rest = rest.replace(/^\([^)]*\)/, "").trim();
  return !rest;
}

function compactAssistantSegments(
  segments: NpcSegment[],
  maxTotal = MAX_DISPLAY_SEGMENTS,
  maxNarrationBlocks = MAX_NARRATION_BLOCKS,
): NpcSegment[] {
  if (!segments.length) return [];

  const merged: NpcSegment[] = [];
  for (const seg of segments) {
    const text = (seg.text || "").trim();
    if (!text) continue;
    const last = merged[merged.length - 1];
    if (last && last.speaker === "내레이션" && seg.speaker === "내레이션") {
      last.text = `${last.text}\n\n${text}`;
    } else {
      merged.push({ speaker: seg.speaker, text });
    }
  }

  const narrIndices = merged
    .map((s, i) => (s.speaker === "내레이션" ? i : -1))
    .filter((i) => i >= 0);
  let compacted = merged;
  if (narrIndices.length > maxNarrationBlocks) {
    const allNarr = narrIndices.map((i) => merged[i].text).join("\n\n");
    const out: NpcSegment[] = [];
    let inserted = false;
    for (let i = 0; i < merged.length; i++) {
      if (merged[i].speaker === "내레이션") {
        if (!inserted && i === narrIndices[0]) {
          out.push({ speaker: "내레이션", text: allNarr });
          inserted = true;
        }
        continue;
      }
      out.push(merged[i]);
    }
    compacted = out;
  }

  if (compacted.length <= maxTotal) return compacted;

  const narrOnly = compacted.filter((s) => s.speaker === "내레이션");
  const npcOnly = compacted.filter((s) => s.speaker !== "내레이션");
  const narrSlot = narrOnly.length ? 1 : 0;
  const keepNpcs = Math.max(1, maxTotal - narrSlot);

  if (npcOnly.length <= keepNpcs) {
    return [...npcOnly, ...narrOnly].slice(0, maxTotal);
  }

  const kept = npcOnly.slice(0, keepNpcs - 1);
  const overflow = npcOnly.slice(keepNpcs - 1);
  kept.push({
    speaker: overflow[0].speaker,
    text: overflow.map((s) => s.text).join("\n\n"),
  });
  if (narrOnly.length) {
    kept.push({ speaker: "내레이션", text: narrOnly[0].text });
  }
  return kept.slice(0, maxTotal);
}

function appendNarration(out: NpcSegment[], block: string): void {
  const last = out[out.length - 1];
  if (last && last.speaker === "내레이션") {
    last.text = `${last.text}\n\n${block}`;
  } else {
    out.push({ speaker: "내레이션", text: block });
  }
}

function splitByUtterances(
  blocks: string[],
  names: string[],
  playerName?: string | null,
): NpcSegment[] {
  const out: NpcSegment[] = [];
  let activeNpc: string | null = null;
  let pendingPrefix = "";

  const flushPrefixToLast = () => {
    if (!pendingPrefix) return;
    const last = out[out.length - 1];
    if (last && last.speaker !== "내레이션") {
      last.text = `${last.text}\n\n${pendingPrefix}`;
    } else if (activeNpc) {
      out.push({ speaker: activeNpc, text: pendingPrefix });
    } else {
      out.push({ speaker: "내레이션", text: pendingPrefix });
    }
    pendingPrefix = "";
  };

  for (const block of blocks) {
    if (!block || HR_LINE.test(block)) continue;

    const first = firstLine(block);
    const headerName = speakerHeaderName(first);
    if (headerName && filterSpeaker(headerName, names, playerName) === null) {
      flushPrefixToLast();
      activeNpc = null;
      appendNarration(out, block);
      continue;
    }

    const named = speakerForBlock(block, names, playerName);

    if (named) {
      activeNpc = named;
      if (isNpcHeaderOnly(block, names)) {
        pendingPrefix = pendingPrefix ? `${pendingPrefix}\n\n${block}` : block;
        continue;
      }
      const text = pendingPrefix ? `${pendingPrefix}\n\n${block}` : block;
      out.push({ speaker: named, text });
      pendingPrefix = "";
      continue;
    }

    if (isActionLine(first) && !blockHasDialogue(block)) {
      pendingPrefix = pendingPrefix ? `${pendingPrefix}\n\n${block}` : block;
      continue;
    }

    if (isDialogueLine(first) || blockHasDialogue(block)) {
      const speaker = activeNpc || "내레이션";
      const text = pendingPrefix ? `${pendingPrefix}\n\n${block}` : block;
      out.push({ speaker, text });
      pendingPrefix = "";
      continue;
    }

    flushPrefixToLast();
    activeNpc = null;
    appendNarration(out, block);
  }

  flushPrefixToLast();
  return out;
}

/**
 * 백엔드 dialogue_split.py 와 동일 — 멘트(따옴표 대사) 단위 카드, 최대 5개.
 */
export function splitAssistantIntoSegments(
  text: string,
  npcNames: string[],
  playerName?: string | null,
): NpcSegment[] {
  const trimmed = normalizeAssistantText(text);
  if (!trimmed) return [];

  const known = npcNames.map((n) => (n || "").trim()).filter(Boolean);
  const names = discoverSpeakerNames(trimmed, known, playerName);

  if (names.length === 0) {
    return [{ speaker: "응답", text: trimmed }];
  }

  const preprocessed = preprocessSpeakerLineBreaks(trimmed);
  const rawBlocks = preprocessed.split(/\n\s*\n+/);
  const blocks = rawBlocks
    .map((b) => b.replace(/^\s+|\s+$/g, ""))
    .filter((b) => b.length > 0 && !HR_LINE.test(b));

  return compactAssistantSegments(splitByUtterances(blocks, names, playerName));
}
